core: add %d specifier for the $CREDENTIALS_DIRECTORY

mrc0mmand opened the pull request:
Resolves: #22549

yuwata commented on src/shared/specifier.c:
```diff
@@ -301,6 +302,27 @@ int specifier_os_image_version(char specifier, const void *data, const char *roo
+
+        assert(ret);
+
+        r = get_credentials_dir(&d);
```
IIUC, this reads values from the `CREDENTIALS_DIRECTORY` environment variable, but at the time the unit is loaded, the env var is not set. Hence, I guess this specifier always provides an empty string `""`.,,

poettering commented:
hmm, this doesn't look right. get_credentials_dir() is something you'd call from service code that reads the env var set by the manager. But the specifier here is something that is resolved in the manager itself, not by the service. Thus it is not the right thing to call.

not that %S, %t, %C and so on all resolve to the *root* where the relevant dirs are placed. The same should apply when we add %c, i.e. the dir we place the per-unit creds dirs in, and not the per-unit creds-dir in full. See build_environment() in src/core/execute.c to see how the credentials dir is currently put together.

or in other words, this should resolve to something like `u->manager->prefix[EXEC_DIRECTORY_RUNTIME]` suffixed with `/credentials`

yuwata commented on src/core/unit-printf.c:
```diff
@@ -148,6 +148,28 @@ static int specifier_special_directory(char specifier, const void *data, const c
+                if (!d)
+                        return -ENOMEM;
+        } else
+                d = strdup("");
```
No oom check?

mrc0mmand replied:
ouch, thanks!

mrc0mmand commented:
Oh well, looks like the ExecContext can be `NULL` (at least when running fuzzers):

```
Assertion '_expr_' failed at src/core/unit-printf.c:158, function int specifier_credentials_dir(char, const void *, const char *, const void *, char **)(). Aborting.
==459== ERROR: libFuzzer: deadly signal
    #0 0x42c714 in __sanitizer_print_stack_trace /src/llvm-project/compiler-rt/lib/ubsan/ubsan_diag_standalone.cpp:31:3
    #1 0x47d8c8 in fuzzer::PrintStackTrace() /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerUtil.cpp:210:5
    #2 0x462d23 in fuzzer::Fuzzer::CrashCallback() /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:233:3
    #3 0x7f22184213bf  (/lib/x86_64-linux-gnu/libpthread.so.0+0x143bf)
    #4 0x7f221823103a in raise (/lib/x86_64-linux-gnu/libc.so.6+0x4303a)
    #5 0x7f2218210858 in abort (/lib/x86_64-linux-gnu/libc.so.6+0x22858)
    #6 0x7f2218c41d44 in log_assert_failed /work/build/../../src/systemd/src/basic/log.c:866:9
    #7 0x7f2219725b67 in specifier_credentials_dir /work/build/../../src/systemd/src/core/unit-printf.c:158:14
    #8 0x7f2218b83923 in specifier_printf /work/build/../../src/systemd/src/shared/specifier.c:69:45
2022-03-15 19:49:56,492 - root - INFO - Fuzzer: fuzz-unit-file. Detected bug.
2022-03-15 19:49:56,493 - root - INFO - Trying to reproduce crash using: /tmp/tmpkzv7t54m/crash-897d87b3cdd36e10ea01e079b4f006b9917dbfc9.
    #9 0x7f2219724937 in unit_full_printf_full /work/build/../../src/systemd/src/core/unit-printf.c:269:16
    #10 0x7f221960b2dd in unit_path_printf /work/build/../../src/systemd/src/core/unit-printf.h:16:16
    #11 0x7f221960b2dd in config_parse_path_spec /work/build/../../src/systemd/src/core/load-fragment.c:2225:13
    #12 0x7f2218a6cd8d in next_assignment /work/build/../../src/systemd/src/shared/conf-parser.c:149:24
    #13 0x7f2218a6cd8d in parse_line /work/build/../../src/systemd/src/shared/conf-parser.c:256:16
    #14 0x7f2218a6bd2c in config_parse /work/build/../../src/systemd/src/shared/conf-parser.c:399:21
    #15 0x42e205 in LLVMFuzzerTestOneInput /work/build/../../src/systemd/src/core/fuzz-unit-file.c:70:16
    #16 0x4642d2 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:611:15
    #17 0x463aba in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:514:3
    #18 0x465924 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::__Fuzzer::vector<fuzzer::SizedFile, std::__Fuzzer::allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:826:7
    #19 0x465b59 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, std::__Fuzzer::allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:857:3
    #20 0x4555d0 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:912:6
    #21 0x47e082 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
    #22 0x7f22182120b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x240b2)
    #23 0x407c1d in _start (build-out/fuzz-unit-file+0x407c1d)
```

so I dropped the asssert and moved the check to the condition below.

I wonder if it would make sense to move the context helpers from [execute.c](https://github.com/systemd/systemd/blob/main/src/core/execute.c#L1418-L1472) to `unit.c` and make them public (or at least make them public in `execute.[ch]`), so we could "prettify" the condition a bit (using `ec && exec_context_has_credentials(ec) ...`.

poettering commented on man/systemd.unit.xml:
```diff
@@ -1975,6 +1975,11 @@
+          <row>
+            <entry><literal>%d</literal></entry>
+            <entry>Credentials directory</entry>
+            <entry>This is the value of the <literal>$CREDENTIALS_DIRECTORY</literal> environment variable if available. See section "Credentials" in <citerefentry><refentrytitle>systemd.exec</refentrytitle><manvolnum>5</manvolnum></citerefentry> for more information.</entry>
```
that's not quite the case. Its the directory *root*, i.e. the unit name has to be suffixed to get the per-unit dir.

mrc0mmand replied:
Is it? We explicitly concatenate it with the unit name in both cases iiuc:

https://github.com/systemd/systemd/blob/c8a4306e63a5387813e6e01164e5303e28dc60dc/src/core/execute.c#L1984

poettering replied:
yes, that's what I am saying: `%d` should map to `p->prefix[EXEC_DIRECTORY_RUNTIME]` suffixed with `/credentials`, but *without* the `u->id` additional suffix.

Thus, i'd call this a "root" directory, because in it you find one subdir for each service.

poettering replied:
or hmm, maybe the way you defined it actually does make more sense.

poettering commented:
Hmm, so I think I agree with your way to define it after all, i.e. not the root, but the dir itself. Given we can have exactly zero or one credential dirs per user it's safe to do it that way. 

However, the conditoinalization on ExecContext settings data must be dropped. First, I don't think the empty string really is better than a path that doesn't exist. i.e. if we really have a service for which people use the new specifier but where credentials are not used, then the error path for handling a nonexisting dir below /run/credentials/ is likely better htan the error path for handling the empty path as dir. 

More importantly though: specifier expansion takes place at the time of parsing. This means the strings specifiers expand to should *not* be dependent on other settings from the same file. Because if they would then suddenly the order of assignment would start to matter. i.e. consider this:

```ini
[Service]
SetCredential=foo:bar
ExecStart=/bin/echo %d
```

vs. 

```ini
[Service]
ExecStart=/bin/echo %d
SetCredential=foo:bar
``` 

these two units should really have the same effect. But in your current patch they won't. Because in the second case we'll parse ExecStart= first, and then do specifier expansion at a moment where we don't know anything about the credentials, thus resolve things to "". But in the first case we'll properly resolve things to /run/credentials/<unitname>.

hence: simplify htis, just drop the conditionalization on the execcontext. It's simpler, more robust and independent of ine order then.

mrc0mmand commented:
> Hmm, so I think I agree with your way to define it after all, i.e. not the root, but the dir itself. Given we can have exactly zero or one credential dirs per user it's safe to do it that way.
> 
> However, the conditoinalization on ExecContext settings data must be dropped. First, I don't think the empty string really is better than a path that doesn't exist. i.e. if we really have a service for which people use the new specifier but where credentials are not used, then the error path for handling a nonexisting dir below /run/credentials/ is likely better htan the error path for handling the empty path as dir.
> 
> More importantly though: specifier expansion takes place at the time of parsing. This means the strings specifiers expand to should _not_ be dependent on other settings from the same file. Because if they would then suddenly the order of assignment would start to matter. i.e. consider this:

Ah, thanks, this makes the ExecContext stuff much clearer.

> 
> ```ini
> [Service]
> SetCredential=foo:bar
> ExecStart=/bin/echo %d
> ```
> 
> vs.
> 
> ```ini
> [Service]
> ExecStart=/bin/echo %d
> SetCredential=foo:bar
> ```
> 
> these two units should really have the same effect. But in your current patch they won't. Because in the second case we'll parse ExecStart= first, and then do specifier expansion at a moment where we don't know anything about the credentials, thus resolve things to "". But in the first case we'll properly resolve things to /run/credentials/.
> 
> hence: simplify htis, just drop the conditionalization on the execcontext. It's simpler, more robust and independent of ine order then.

Agreed. Again, thanks for the hints!
